// nebula.gl/dist/lib/feature.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Feature =
/*#__PURE__*/
function () {
  // geo json coordinates
  function Feature(geoJson, style) {
    var original = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    var metadata = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};

    _classCallCheck(this, Feature);

    _defineProperty(this, "geoJson", void 0);

    _defineProperty(this, "style", void 0);

    _defineProperty(this, "original", void 0);

    _defineProperty(this, "metadata", void 0);

    this.geoJson = geoJson;
    this.style = style;
    this.original = original;
    this.metadata = metadata;
  }

  _createClass(Feature, [{
    key: "getCoords",
    value: function getCoords() {
      return this.geoJson.geometry.coordinates;
    }
  }]);

  return Feature;
}();

exports.default = Feature;
