// @nebula.gl/layers/dist/mode-handlers/composite-mode-handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositeModeHandler = void 0;

var _modeHandler = require("./mode-handler.js");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var CompositeModeHandler =
/*#__PURE__*/
function (_ModeHandler) {
  _inherits(CompositeModeHandler, _ModeHandler);

  function CompositeModeHandler(handlers) {
    var _this;

    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, CompositeModeHandler);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CompositeModeHandler).call(this));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handlers", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "options", void 0);

    _this.handlers = handlers;
    _this.options = options;
    return _this;
  }

  _createClass(CompositeModeHandler, [{
    key: "_coalesce",
    value: function _coalesce(callback) {
      var resultEval = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
      var result;

      for (var i = 0; i < this.handlers.length; i++) {
        result = callback(this.handlers[i]);

        if (resultEval ? resultEval(result) : result) {
          break;
        }
      }

      return result;
    }
  }, {
    key: "setFeatureCollection",
    value: function setFeatureCollection(featureCollection) {
      this.handlers.forEach(function (handler) {
        return handler.setFeatureCollection(featureCollection);
      });
    }
  }, {
    key: "setModeConfig",
    value: function setModeConfig(modeConfig) {
      this.handlers.forEach(function (handler) {
        return handler.setModeConfig(modeConfig);
      });
    }
  }, {
    key: "setSelectedFeatureIndexes",
    value: function setSelectedFeatureIndexes(indexes) {
      this.handlers.forEach(function (handler) {
        return handler.setSelectedFeatureIndexes(indexes);
      });
    }
  }, {
    key: "setDeckGlContext",
    value: function setDeckGlContext(context) {
      this.handlers.forEach(function (handler) {
        return handler.setDeckGlContext(context);
      });
    }
  }, {
    key: "handleClick",
    value: function handleClick(event) {
      return this._coalesce(function (handler) {
        return handler.handleClick(event);
      });
    }
  }, {
    key: "handlePointerMove",
    value: function handlePointerMove(event) {
      return this._coalesce(function (handler) {
        return handler.handlePointerMove(event);
      }, function (result) {
        return result && Boolean(result.editAction);
      });
    }
  }, {
    key: "handleStartDragging",
    value: function handleStartDragging(event) {
      return this._coalesce(function (handler) {
        return handler.handleStartDragging(event);
      });
    }
  }, {
    key: "handleStopDragging",
    value: function handleStopDragging(event) {
      return this._coalesce(function (handler) {
        return handler.handleStopDragging(event);
      });
    }
  }, {
    key: "getTentativeFeature",
    value: function getTentativeFeature() {
      return this._coalesce(function (handler) {
        return handler.getTentativeFeature();
      });
    }
  }, {
    key: "getEditHandles",
    value: function getEditHandles(picks, groundCoords) {
      // TODO: Combine the handles *BUT* make sure if none of the results have
      // changed to return the same object so that "editHandles !== this.state.editHandles"
      // in editable-geojson-layer works.
      return this._coalesce(function (handler) {
        return handler.getEditHandles(picks, groundCoords);
      }, function (handles) {
        return Array.isArray(handles) && handles.length > 0;
      });
    }
  }, {
    key: "getCursor",
    value: function getCursor(_ref) {
      var isDragging = _ref.isDragging;
      return this._coalesce(function (handler) {
        return handler.getCursor({
          isDragging: isDragging
        });
      });
    }
  }]);

  return CompositeModeHandler;
}(_modeHandler.ModeHandler);

exports.CompositeModeHandler = CompositeModeHandler;
